pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -241,28 +241,23 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(env *model.Env
 		}
 
 		for vsvcHost, gatewayHost := range matchingHosts {
-			_, f := vhostDomains[vsvcHost]
-			if f {
-				// RDS would reject this, resulting in all vhosts rejection.
-				push.Add(model.DuplicatedDomains, vsvcHost, node,
-					fmt.Sprintf("%s duplicate domain %s for %s", node.ID, vsvcHost, v.Name))
-				continue
-			}
-			vhostDomains[vsvcHost] = true
-			host := route.VirtualHost{
-				Name:    fmt.Sprintf("%s:%d", v.Name, port),
-				Domains: []string{vsvcHost, fmt.Sprintf("%s:%d", vsvcHost, port)},
-				Routes:  routes,
-			}
-
-			if tlsRedirect[gatewayHost] {
-				host.RequireTls = route.VirtualHost_ALL
+			if currentVhost, exists := vHostDedupMap[vsvcHost]; exists {
+				currentVhost.Routes = append(currentVhost.Routes, routes...)
+			} else {
+				newVhost := &route.VirtualHost{
+					Name:    fmt.Sprintf("%s:%d", vsvcHost, port),
+					Domains: []string{vsvcHost, fmt.Sprintf("%s:%d", vsvcHost, port)},
+					Routes:  routes,
+				}
+				if tlsRedirect[gatewayHost] {
+					newVhost.RequireTls = route.VirtualHost_ALL
+				}
+				vHostDedupMap[vsvcHost] = newVhost
 			}
-			virtualHosts = append(virtualHosts, host)
 		}
 	}
 
-	if len(virtualHosts) == 0 {
+	if len(vHostDedupMap) == 0 {
```

Comment:
nit: move `virtualHosts := make([]route.VirtualHost, 0, len(virtualServices))` down here to line 259 so it is clear we don't use it until we convert `vHostDedupMap`?